### ISSUE TITLE
How should images be preprocessed for forward pass during AdaBN?

### ISSUE THREAD

**Issue by shivani21998** [ultralytics/ultralytics#20767]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I'm applying AdaBN (Adaptive Batch Normalization) to adapt a trained YOLOv8 model to a new domain using unlabelled images. I'm running forward passes through the model to update the BN running stats like this:

```python
with torch.no_grad():
    for imgs in loader:
        imgs = imgs.to(device)
        model(imgs)
```
To prepare the data, I'm currently preprocessing images as follows:

Using letterbox() to resize and pad to 640×640

Converting image format from RGB → BGR → RGB (to align with OpenCV + PyTorch)

Transposing HWC → CHW

Normalizing pixel values to [0, 1]

Converting to PyTorch tensors
```python
img = np.array(Image.open(img_path).convert("RGB"))
img_bgr = img[..., ::-1].copy()  # RGB to BGR
img_resized = letterbox(img_bgr, new_shape=(640, 640))
img_rgb = img_resized[..., ::-1].transpose(2, 0, 1)  # BGR to RGB, HWC to CHW
img_tensor = torch.from_numpy(np.ascontiguousarray(img_rgb)).float() / 255.0
```

My question: Is this the correct preprocessing workflow for images used in forward passes during AdaBN

### Additional

Issue: 
After applying AdaBN with the above preprocessing steps, I observed that the results on my custom dataset (which is in the same domain as the test set) underperformed compared to the original YOLOv8 model without adaptation.

I'm trying to determine whether this is due to incorrect image preprocessing or something else.

Any insights would be greatly appreciated. Thank you for the amazing work on YOLOv8!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @shivani21998, thank you for your detailed question and for using Ultralytics 🚀! For many preprocessing and model usage questions, we recommend reviewing the [Docs](https://docs.ultralytics.com/) where [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples can help clarify best practices.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate.

For custom training or adaptation ❓ questions like yours, please include as much information as possible, such as example images, your full preprocessing code, and any relevant logs or metrics. Also, double-check that you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

You’re welcome to join our community for support and discussion:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time help
- [Discourse](https://community.ultralytics.com/) for in-depth conversations
- [Subreddit](https://reddit.com/r/Ultralytics) to connect with other users

## Upgrade

Make sure you're running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to rule out already-resolved issues:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests ensure all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) run correctly on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help you get started quickly. An Ultralytics engineer will review your question and assist you further soon! 😊

**Comment by glenn-jocher** [comment]
@shivani21998 Your preprocessing approach is mostly correct, but there's a redundant step that could be causing issues. You're converting RGB → BGR → RGB, which is unnecessary since YOLO11 handles the BGR to RGB conversion internally during preprocessing.

For AdaBN forward passes, you should use the same preprocessing pipeline as the model's inference. The simplest approach is to leverage the model's built-in preprocessing by using the [preprocess method](https://docs.ultralytics.com/reference/engine/predictor/#preprocess) directly:

```python
from ultralytics.data.augment import LetterBox

# Use model's preprocessing pipeline
letterbox = LetterBox(new_shape=(640, 640), auto=False, stride=32)
img = np.array(Image.open(img_path).convert("RGB"))
processed = letterbox(image=img)
img_tensor = torch.from_numpy(processed).float().unsqueeze(0).to(device) / 255.0

with torch.no_grad():
    model(img_tensor)
```

The underperformance you're seeing could be due to the extra BGR conversion step or domain mismatch in your preprocessing pipeline compared to the original training data preprocessing.

**Comment by san-kim0106** [comment]
Hi there @shivani21998,

I leave a short message to ask if you've observed any notable improvements in unseen domains when applying AdaBN to a standard YOLO model pre-trained on COCO?”

Thanks
